README
ibus-braille
====================

<b>Braille input method for ibus </b>


ibus-braille is an ibus input engine based on six key approach of braille. We express our gratitude to Indic project & Swathanthra Malayalam Computing(SMC) for mentoring this project. We consider the acceptance of this project by Swathanthra Malayalam Computing and Google as a new flowering of the effort of louis braille. By default it uses letters f, d, s, j, k, and l in the keyboard to represent 1, 2, 3, 4, 5 and 6 of the braille dots respectively. One can use different combinations of braille to produce text. For example key 'f' will produce 'a', 'f' and 'd' will produce 'b' and 'f' and 'j' will produce 'c' and like. The combination of keys should be released together fafter pressing them.

	
<b>Languages : -</b> This version comes with  seven languages English, Malayalam, Hindi, Kannada, Tamil, French, and Spanish.
English is the default language, and one can switch to other languages by pressing language switch key (Pause Break). one can go back to previus language using the same key. One can select the required languages by checking them in general page of ibus-braille-preferences. Also default language can be change by using the combo box that appears in the same page.


<b>Contractions : -</b> All the grade 2 and grade 3 contractions in English and Malayalam is available in ibus-braille. 
In order to activate contractions one should uncheck the simple mode. To produce contraction, press the combination of
letters after pressing the contraction key, Dot 5, 5-6, 4-5-6, 4, 3, 3-6, 6, 3, 1-5-6 are the contraction signs.
For example, press 'f' and 'd' together after pressing 'j' will produce word "better" and like. Along with these contractions,
we have provided some contractions in the simple mode itself. For this purpose we have used the combinations which
are not used for alphabets.

<b>Numerals : -</b> One can select numeral as language. Nemeth code approach is accepted here. In this mode, 
d, s, k, l representing 2, 3, 5, 6 of the braille dots are used for numbers that is d for 1 and d, 
s together for 2 etc. Other letters are written as in braille mode, f, k, l combination is used to produce 
underscore and this combination can be used  to produce line in any text. Numeral mode is used for handling 
mathematics.

<b>Punctuations : -</b> To produce any punctuation one should press semicolon in the ordinary keyboard and then press necessary combinations.


<b>Abbreviations : -</b> letter 'a' is the abbreviation key and one should press 'a'  after typing the abbreviation. For example pressing 'a' 
after typing 'ab' will expand the word 'ab' to 'about' This facility will increase the speed of producing text.

<b>Abbreviation-Editor : -</b> It is a unique feature of isb which enable a user to create whatever abbreviation he/she requires. 
In order to edit the abbreviation one has to open ibus-braille-abbreviation-editor then select the language of which abbreviation to be edited. 
For adding a new abbreviation one has to click add button which will prompt a dialog with two entries: abbreviation and expantion, user has to fill apropriate data and click add button. One can remove an abbreviation by clicking on remove button. One has to press save button to make it effective. Abbreviations thus saved can be shared with others using the export button which will prompt a save dialog.
Restore button will delete all newly added abbreviations and restore the list to the default position.
For english we have already provided all the grade 3 abbreviations in the package. 




<b>Beginning list and middle list : -</b> One can switch between middle list and beginning list using map-switching key(left ALT). 
This will help the user to write any letter anywhere. For example  one can write a vowel in hindi inside a word. 
To write a full vowel instead of a sign, one should press left alt and then type the combination needed for the vowel.

<b>Simple-Mode : -</b> ibus braille is installed as the simple mode unchecked. This will be complicated for a beginner. To avoid these complications a newbie can disable abbreviations and contractions by checking simple mode from general page of ibus-braille-preferences.

<b>Text Manipulation : -</b> capitol/chill key('g') is used to produce capital letter in English and chillu in malayalam. 
pressing this key before typing the braille combination to make a letter capital and pressing this key after typing 
the braille combination to make a letter chillu in malayalam. If this capitol key is pressed before using a contracted 
word, whole letters will be in capital. Letter deletion key('h') will delete the last letter typed and pressing letter deletion key and capitol/chill key together will delete the last word. 
<b>Caps lock : -</b> One can enable and disable caps lock by pressing capitol/chill key two time continuesly. 


<b>Key-remapping : -</b> One can change the all keys by using ket/shortcut page of ibus-braille-preferences. Users having non qwerty keyboard can use this to setup six appropriate keys on there keyboard. Here user can also change the key for language switching,map switching and capitol/chill as well. By default the following keys are given for different functionalities:
f, d, s, j, k, l for dots 1, 2, 3, 4, 5, 6 of the braille dots.  Semicolon as punctuation key, 'g' as capital/chilli, 'h' for deletion of last letter, left alt map switching key,
'a' as abbreviation expansion key. User can tab to listen to the key assigned to the function and simply press the needed key to change it. After making the changes press apply button to effect the changes. Here also one can use restore button to come back to the default settings.

<b>ibus-braille language editor : -</b> The language editor enable the user to add a new language, remove the existing language, add a new map 
of contractions, remove the map of contractions, and manipulate the characters and contractions of existing languages.
One can also share the languages, maps, and contractions he has created, with others using the export and import facility. 


one cant start the process of editing by selecting the languge to be edited from combobox. This is the list of existing languages and one can press add-language button for creating new language. After pressing this button, one should enter the name of the language with its e-speak variant name putting a hyphen between them.
Now one can work upon the language by selecting it from the list. Using remove-language button one can remove any language.

<b>Editing Map : -</b> A map is an arrangement for producing a character or list or a contraction list. 
A map is a key combination which does not produce any character but enable the user to produce the list of characters and contractions. 
After pressing the add-map button it will ask for the entry. One can press the braille combination using and press the add button. If one has pressed j and l the editor will show "46" and in the 
process map for 4, 6 will appear in the editor and one can add any contractions in the 4, 6 list.


<b>Editing Beginning list : -</b> This page list out all the characters that should appear after pressing the space. For 
example it contains all the vowels and other alphabets of  indian languages, and all the alphabets and 
contractions given in the braille six key combinations in english. Let us now see how one can change 
or remove a character. Just tab and then a list will appear which contains the number combinations are 
given. Again tab and add new row button will appear. After pressing the button, editor will ask for the 
braille combination and enter it using six key combinations as done in map creation. Again tab and you 
will get the column for adding the value. Then again tab and press add button. If one needs further entry 
add row button will appear again.

<b>Editing middle list : -</b> In the map page and curser is at beginning, use right arrow to select the middle list. Middle list contains 
all the characters that should appear when a character  is already in position. It contains all the 
signs attached to the vowels and other alphabets in indian languages and all the alphabets and middle 
contractions like cc dd etc. 

<b>Editing Punctuation list : -</b> press right arrow and one can go to punctuation list  and here one can see the list 
of combinations for punctuations which will appear after pressing the punctution key(semicolon) in the ordinary key 
board.

please note that one can't remove beginning,middle and pucntation lists because they are essential for the working of a language.
Using the right arrow one can manipulate other maps such as 4 6, 5, 4 5 6, etc in english. Again one 
can remove any unwanted row as well. One can clear all the entries using clear-all button, After the operations one should save/save-all and quit the editor to bring the changes in 
to effect.


<b>Accessibility : -</b> ibus-shrada-braille is fully orca supported and one can use the orca preferences for changing voices, punctuation levels 
to be pronounced, key echo  by character etc. In order to activate orca preferences one should 
press insert+space together. After the installation of the program,  
we recommend you to open the orca preferences and check the echo by character and uncheck the 
key echo in the key echo page to get a better speech support. To get speech support for Kannada and Tamil one should change 
the language to that language using the orca preferences.


Copyright (c) 2014-2015 IBus-Braille Development Team 

See COPYING file for License

 


debian/changelog
ibus-braille (1.1-1) focal; urgency=medium

  * Bug fixes

 -- Nalin <nalin.x.linux@gmail.com>  Sat, 09 Jul 2022 09:11:59 +0530

ibus-braille (1.0-1) focal; urgency=medium

  * New version with libbraille-input

 -- Nalin <nalin.x.linux@gmail.com>  Tue, 24 May 2022 12:04:18 +0530

ibus-braille (0.6.2) focal; urgency=medium

  * Dependency back translation table version corrected

 -- Nalin <nalin.x.linux@gmail.com>  Mon, 05 Apr 2021 19:01:08 +0530

ibus-braille (0.6.1) focal; urgency=medium

  * Liblouis path fixed
  * Liblouis back translation table installation path fixed

 -- Nalin <nalin.x.linux@gmail.com>  Sun, 04 Apr 2021 19:02:05 +0530

ibus-braille (0.6) focal; urgency=medium

  * Liblouis integration - TTS synthesizer language switching enabled
  * Liblouis integration - Abbreviations enabled and made more simple
  * Liblouis integration - Text announcement improved
  * Configuration directory changed to ~/.ibus-braille
  * Icon updated
  * Makefile.am required file 'README' added

 -- Nalin <nalin.x.linux@gmail.com>  Sat, 27 Mar 2021 08:43:49 +0530

ibus-braille (0.5) focal; urgency=medium

  * liblouis-back-translation-tables added

 -- Nalin <nalin.x.linux@gmail.com>  Fri, 12 Feb 2021 19:38:34 +0530

ibus-braille (0.4-3) focal; urgency=medium

  * Built-in engine fixed

 -- Nalin <nalin.x.linux@gmail.com>  Tue, 26 Jan 2021 10:20:25 +0530

ibus-braille (0.4-2) focal; urgency=medium

  * Liblouis integration improved
  * TTS announcement delay removed
  * Bugfixes

 -- Nalin <nalin.x.linux@gmail.com>  Sat, 24 Oct 2020 09:09:33 +0530

ibus-braille (0.3-5) unstable; urgency=medium

  * control: Drop python2 dependencies, add missing gir1.2-pango-1.0 dependency.
  * rules: Use --with python3 instead of python2.

 -- Samuel Thibault <sthibault@debian.org>  Mon, 22 Jul 2019 21:07:07 +0200

ibus-braille (0.3-4) unstable; urgency=medium

  * Team upload.

  [ Samuel Thibault ]
  * control: Bump Standards-Version to 4.2.1 (no changes).

  [ Boyuan Yang ]
  * control: Update Vcs-* fields and use the Salsa input-method-team repo.
  * copyright: Fix format to eliminate lintian warnings.

 -- Boyuan Yang <byang@debian.org>  Thu, 25 Oct 2018 00:39:30 -0400

ibus-braille (0.3-3) unstable; urgency=medium

  * control: Fix maintainer mailing list address (Closes: Bug#899540)

 -- Samuel Thibault <sthibault@debian.org>  Thu, 24 May 2018 09:59:21 +0200

ibus-braille (0.3-2) unstable; urgency=medium

  * Use canonical anonscm vcs URL.
  * control: Migrate priority to optional.
  * control: Bump Standards-Version to 4.1.4.
  * control: Remove old ibus-sharada-braille transitional package.

 -- Samuel Thibault <sthibault@debian.org>  Sat, 28 Apr 2018 17:26:52 +0200

ibus-braille (0.3-1) unstable; urgency=medium

  * New upstream release.
    - control: Depend on python3-louis.
    - patches/timestamp: Comment, fixed upstream.
  * control: Bump Standards-Version to 3.9.8 (no change)
  * watch: Add.

 -- Samuel Thibault <sthibault@debian.org>  Thu, 13 Oct 2016 10:32:16 +0200

ibus-braille (0.1.2.99+git1.a95477d-4) unstable; urgency=medium

  * patches/timestamp: Fix unreproducibility due to timestamp in version number.
  * control: Bump Standards-Version to 3.9.7 (no change).

 -- Samuel Thibault <sthibault@debian.org>  Tue, 23 Feb 2016 01:04:09 +0100

ibus-braille (0.1.2.99+git1.a95477d-3) unstable; urgency=medium

  * control: ibus-braille: Turn arch: all, since it does not depend on the
    arch any more.

 -- Samuel Thibault <sthibault@debian.org>  Tue, 29 Dec 2015 17:56:58 +0100

ibus-braille (0.1.2.99+git1.a95477d-2) unstable; urgency=medium

  * rules: Fix installing files... and switch to python3
  * control: Recommend python3-speechd instead of python-speechd (which was
    removed).

 -- Samuel Thibault <sthibault@debian.org>  Tue, 29 Dec 2015 15:25:42 +0100

ibus-braille (0.1.2.99+git1.a95477d-1) unstable; urgency=medium

  * New upstream snapshot
    - rename packages from ibus-sharada-braille to ibus-braille.
    - keep an ibus-sharada-braille transitional package.
  * Bump Standards-Version to 3.9.6 (no changes).
  * control: Add missing python-configparser dependency.

 -- Samuel Thibault <sthibault@debian.org>  Mon, 28 Sep 2015 00:28:10 +0200

ibus-sharada-braille (0.1.2.99+git0.c90ffb02-2) unstable; urgency=medium

  * copyright: Fix licence of packaging/ibus-sharada-braille.spec.

 -- Samuel Thibault <sthibault@debian.org>  Mon, 15 Sep 2014 00:35:32 +0200

ibus-sharada-braille (0.1.2.99+git0.c90ffb02-1) unstable; urgency=low

  * Initial release (Closes: #759251)

 -- Samuel Thibault <sthibault@debian.org>  Fri, 29 Aug 2014 23:24:10 +0200


IBusBraille/preferences.py
###########################################################################
#    IBus-Braille - Braille input engine for IBus
#
#    Copyright (C) 2014-2015 Nalin Sathyan <nalin.x.linux@gmail.com>
#    Copyright (C) 2016-2017 Anwar N <anwar3746@gmail.com>
#    Copyright (C) 2022-2023 Nalin Sathyan <nalin.x.linux@gmail.com>
#    
#    This project was developed twice under Google Summer of Code program 
#    GSoC 2014 and 2016 under the guidance of Samuel Thibault.
#
#    Entire project is rewritten by Nalin Sathyan and powered by Zendalona
#    wwww.zendalona.com
#
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
###########################################################################

import sys
import configparser

import gi

gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
from gi.repository import Gdk

from IBusBraille import global_var

from IBusBraille import localization
_ = localization._

from brailleinput.engine import Keys


class Preferences():
	
	default_keycode_map = {Keys.Dot1 : 33, Keys.Dot2 : 32, Keys.Dot3 : 31,
		Keys.Dot4 : 36, Keys.Dot5 : 37, Keys.Dot6 : 38, Keys.Dot7 : 44,
		Keys.Dot8 : 52, Keys.ToggleBrailleMode : 88, Keys.Abbreviation : 30, Keys.Capital : 34,
		Keys.LetterDeletion : 35, Keys.Punctuation : 39, Keys.BegMidSwitch : 23,
		Keys.OneHandSkip : 20, Keys.Space : 57, Keys.NewLine : 28, Keys.NewLine2 : 49,
		Keys.LangSwitch : 13, Keys.BuiltInSel : 54, Keys.L1 : 2, Keys.L2 : 3,
		Keys.L3 : 4, Keys.L4 : 5, Keys.L5 : 6, Keys.L6 : 7, Keys.L7 : 8, Keys.L8 : 9,
		Keys.L9 : 10, Keys.L10 : 11, Keys.L11 : 12, Keys.L12 : 13}

	
	def __init__(self):
		self.set_default_preferences();
		
	def get_keycode_map(self):
		return {v: k for k, v in self.keycode_map.items()}


	def set_default_preferences(self):
		self.checked_languages_built_in = ["english","hindi", "kannada", "malayalam", "spanish", "tamil", "numerical", "arabic", "english-grade-2", "sanskrit", "braille-patterns", "french",]
		self.checked_languages_liblouis = ["English-US-Grade-1", "English-US-Grade-2", "Malayalam-Grade-1", "Malayalam-Grade-2", "Hindi-Grade-1", 
		"Spanish", "Portuguese-Grade-1", "Arabic-Grade-1", "Nemeth-Code-for-Mathematics", "Russian-Grade-1", "English-US-Grade-3", "Malayalam-Grade-3" ]
		
		self.keycode_map = self.default_keycode_map.copy();

		self.default_liblouis_mode = True;
		self.auto_capitalize_sentence =  True
		self.auto_capitalize_line =  False
		self.simple_mode =  False
		self.speech =  True
		self.conventional_braille = False;
		self.one_hand_mode = False
		self.one_hand_conversion_delay = 500
		self.line_limit =  100
		self.auto_new_line = 1;


	
	def load_preferences_from_file(self, filename):
		try:
			cp = configparser.ConfigParser()
			print(filename)
			cp.read(filename)
			self.checked_languages_built_in = cp.get('language',"checked-languages-built-in").split(",")
			self.checked_languages_liblouis = cp.get('language',"checked-languages-liblouis").split(",")
			self.default_liblouis_mode = int(cp.get('language',"default-liblouis-mode"))

			for item in Keys:
				self.keycode_map[item] = int(cp.get('key-bindings',str(item)))
			
			self.auto_capitalize_sentence = int(cp.get('auto-capitalizing',"auto-capitalize-sentence"))
			self.auto_capitalize_line = int(cp.get('auto-capitalizing',"auto-capitalize-line"))

			self.simple_mode = int(cp.get('built-in',"simple-mode"))
			self.conventional_braille = int(cp.get('built-in',"conventional-braille"))

			self.line_limit = int(cp.get('line-limiting','line-limit'))
			self.auto_new_line = int(cp.get('line-limiting','auto-new-line'))
			
			self.speech = int(cp.get('voice-announcement',"speech"))

			self.one_hand_mode = int(cp.get('one-hand-typing',"one-hand-mode"))
			self.one_hand_conversion_delay = int(cp.get('one-hand-typing',"one-hand-conversion-delay"));
			
		except:
			print("Configuration reading error : ", sys.exc_info()[0])
			self.set_default_preferences()

	def save_preferences_to_file(self, filename):
		
		cp = configparser.ConfigParser()
		cp.add_section('language')
		cp.add_section('built-in')
		cp.add_section('auto-capitalizing')
		cp.add_section('line-limiting')
		cp.add_section('voice-announcement')
		cp.add_section('one-hand-typing')
		cp.add_section('key-bindings')

		cp.set('language',"checked-languages-built-in",",".join(self.checked_languages_built_in))
		cp.set('language',"checked-languages-liblouis",",".join(self.checked_languages_liblouis))
		cp.set('language',"default-liblouis-mode",str(int(self.default_liblouis_mode)))

		for key in self.keycode_map.keys():
			cp.set('key-bindings',str(key),str(self.keycode_map[key]))

		cp.set('auto-capitalizing',"auto-capitalize-sentence",str(int(self.auto_capitalize_sentence)))
		cp.set('auto-capitalizing',"auto-capitalize-line",str(int(self.auto_capitalize_line)))

		cp.set('built-in',"simple-mode",str(int(self.simple_mode)))
		cp.set('built-in',"conventional-braille",str(int(self.conventional_braille)))

		cp.set('line-limiting',"line-limit",str(int(self.line_limit)))
		cp.set('line-limiting',"auto-new-line",str(int(self.auto_new_line)))

		cp.set('voice-announcement',"speech",str(int(self.speech)))

		cp.set('one-hand-typing',"one-hand-mode",str(int(self.one_hand_mode)))
		cp.set('one-hand-typing',"one-hand-conversion-delay",str(int(self.one_hand_conversion_delay)))

		with open(filename , 'w') as configfile:
			cp.write(configfile)
			
	def set_built_in_language_list(self, list_):
		self.list_of_available_built_in_languages = list_

	def set_liblouis_language_list(self, list_):
		self.list_of_available_liblouis_languages = list_
		
			
	def open_preferences_dialog(self):
		self.guibuilder = Gtk.Builder()
		self.guibuilder.add_from_file(global_var.data_dir+"preferences.glade")
		self.window = self.guibuilder.get_object("window")


		# Setting Liblouis language comboboxes
		self.box_liblouis = self.guibuilder.get_object("box_liblouis")

		liststore_liblouis_languages = Gtk.ListStore(str)
		for item in self.list_of_available_liblouis_languages:
			liststore_liblouis_languages.append([item]);
		
		self.list_of_liblouis_language_comboboxes = []
		for i in range (0, 12) :		
			# Setting combobox 
			combo = Gtk.ComboBox.new_with_model(liststore_liblouis_languages)
			renderer_text = Gtk.CellRendererText()
			combo.pack_start(renderer_text, True)
			combo.add_attribute(renderer_text, "text", 0)
			
			cur_lang = self.checked_languages_liblouis[i]
			cur_lang_index =  self.list_of_available_liblouis_languages.index(cur_lang)
			combo.set_active(cur_lang_index)
			combo.show()
			
			# Setting label
			label = Gtk.Label(_("Table ")+str(i+1))
			label.set_mnemonic_widget(combo)
			label.show()
			
			# Packing label and combobox
			box = Gtk.HBox()
			box.pack_start(label,True,True,10);
			box.pack_start(combo,True,True,10);
			box.show()
			
			self.box_liblouis.pack_start(box,False,False,10)
			
			self.list_of_liblouis_language_comboboxes.append(combo)


		# Setting Built-in language comboboxes
		self.box_built_in = self.guibuilder.get_object("box_built_in")

		liststore_built_in_languages = Gtk.ListStore(str)
		for item in self.list_of_available_built_in_languages:
			liststore_built_in_languages.append([item]);
		
		self.list_of_built_in_language_comboboxes = []
		for i in range (0, 12) :		
			# Setting combobox 
			combo = Gtk.ComboBox.new_with_model(liststore_built_in_languages)
			renderer_text = Gtk.CellRendererText()
			combo.pack_start(renderer_text, True)
			combo.add_attribute(renderer_text, "text", 0)
			
			cur_lang = self.checked_languages_built_in[i]
			cur_lang_index =  self.list_of_available_built_in_languages.index(cur_lang)
			combo.set_active(cur_lang_index)
			combo.show()
			
			# Setting label
			label = Gtk.Label(_("Language "+str(i+1)))
			label.set_mnemonic_widget(combo)
			label.show()
			
			# Packing label and combobox
			box = Gtk.HBox()
			box.pack_start(label,True,True,20);
			box.pack_start(combo,True,True,20);
			box.show()			
			
			self.box_built_in.pack_start(box,False,False,10);
			
			self.list_of_built_in_language_comboboxes.append(combo)
		

		# Set key/shortcuts in UI
		self.keycode_map_new = self.keycode_map.copy()
		for item in Keys:
			widget = self.guibuilder.get_object(str(item).replace("Keys.",""))
			hardware_keycode = self.keycode_map_new[item]
			
			keymap = Gdk.Keymap.get_default()
			entries_for_keycode = keymap.get_entries_for_keycode(hardware_keycode + 8)
			entries = entries_for_keycode[-1]
			text = Gdk.keyval_name(entries[0])
			
			widget.set_text(text)
		
		
		#Set auto capitalize sentence checkbox
		self.checkbutton_auto_capitalize_sentence = self.guibuilder.get_object("checkbutton_auto_capitalize_sentence")
		self.checkbutton_auto_capitalize_sentence.set_active(self.auto_capitalize_sentence)

		#Set auto capitalize line checkbox
		self.checkbutton_auto_capitalize_line = self.guibuilder.get_object("checkbutton_auto_capitalize_line")
		self.checkbutton_auto_capitalize_line.set_active(self.auto_capitalize_line)

		#Set Simple mode checkbox
		self.checkbutton_simple_mode = self.guibuilder.get_object("checkbutton_simple_mode")
		self.checkbutton_simple_mode.set_active(self.simple_mode)

		#Set conventional-braille checkbox
		self.checkbutton_conventional_braille = self.guibuilder.get_object("checkbutton_conventional_braille")
		self.checkbutton_conventional_braille.set_active(self.conventional_braille)

		#Set default TableType combobox
		self.combobox_table_type = self.guibuilder.get_object("combobox_table_type")
		self.combobox_table_type.set_active(self.default_liblouis_mode)

		
		#Set auto new line checkbox
		self.checkbutton_auto_new_line = self.guibuilder.get_object("checkbutton_auto_new_line")
		self.checkbutton_auto_new_line.set_active(self.auto_new_line)

		#Set Speech checkbox
		self.checkbutton_speech = self.guibuilder.get_object("checkbutton_speech")
		self.checkbutton_speech.set_active(self.speech)

		#Set auto new line checkbox
		self.spinbutton_line_limit = self.guibuilder.get_object("spinbutton_line_limit")
		self.spinbutton_line_limit.set_value(self.line_limit)


		#Set one-hand-mode checkbox
		self.checkbutton_one_hand_mode = self.guibuilder.get_object("checkbutton_one_hand_mode")
		self.checkbutton_one_hand_mode.set_active(self.one_hand_mode)

		#Set one-hand-mode conversion delay
		self.scale_one_hand_conversion_delay = self.guibuilder.get_object("scale_one_hand_conversion_delay")
		self.scale_one_hand_conversion_delay.set_value(self.one_hand_conversion_delay)			
		
		self.guibuilder.connect_signals(self)
		self.window.show()



	def on_key_press(self,widget,event):
		hardware_keycode = int(event.hardware_keycode)
		hardware_keycode_ibus = hardware_keycode - 8 
		print(hardware_keycode_ibus)
		# Avoid essential keys like Esc, Tab, Shift, R-Shift, Left, Right, Up, Down 
		if (hardware_keycode not in [9,23,50,62,113,114,111,116]):
			widget_name = Gtk.Buildable.get_name(widget)	
			key_in_widget = Keys.__dict__[widget_name]
			if self.keycode_map_new[key_in_widget] != hardware_keycode_ibus:
				if hardware_keycode_ibus in self.keycode_map_new.values():
					widget.set_text("None")
					self.keycode_map_new[key_in_widget] = None;
				else:
					self.keycode_map_new[key_in_widget] = hardware_keycode_ibus
					keymap = Gdk.Keymap.get_default()
					entries_for_keycode = keymap.get_entries_for_keycode(hardware_keycode)
					entries = entries_for_keycode[-1]
					text = Gdk.keyval_name(entries[0])
					widget.set_text(text)

	def set_on_apply_preferences_callback(self, function):
		self.apply_preferences = function
	
	def on_preferences_dialog_apply(self, widget):
		# Copy new key bindings if there is no unassigned keybinding
		if (None in self.keycode_map_new.values()):
			dialog = Gtk.MessageDialog( flags=0, message_type=Gtk.MessageType.INFO,
			buttons=Gtk.ButtonsType.OK, text=_("Unassigned keybinding found"))
			dialog.format_secondary_text(_("Find and fix unassigned keybindings before applying"))
			dialog.run()
			dialog.destroy()
			return
		else:
			self.keycode_map = self.keycode_map_new.copy()

		self.checked_languages_liblouis = []
		for combo in self.list_of_liblouis_language_comboboxes:
			index = combo.get_active()
			self.checked_languages_liblouis.append(self.list_of_available_liblouis_languages[index])
		print(self.checked_languages_liblouis)

		self.checked_languages_built_in = []
		for combo in self.list_of_built_in_language_comboboxes:
			index = combo.get_active()
			self.checked_languages_built_in.append(self.list_of_available_built_in_languages[index])
		print(self.checked_languages_built_in)
		
		# Copy new values
		self.auto_capitalize_sentence = self.checkbutton_auto_capitalize_sentence.get_active()
		self.auto_capitalize_line = self.checkbutton_auto_capitalize_line.get_active()
		self.simple_mode = self.checkbutton_simple_mode.get_active()
		self.conventional_braille = self.checkbutton_conventional_braille.get_active()
		self.default_liblouis_mode = self.combobox_table_type.get_active()
		
		self.one_hand_mode = self.checkbutton_one_hand_mode.get_active()
		self.one_hand_conversion_delay = self.scale_one_hand_conversion_delay.get_value()

		self.auto_new_line = self.checkbutton_auto_new_line.get_active()
		self.line_limit = self.spinbutton_line_limit.get_value_as_int()

		self.speech = self.checkbutton_speech.get_active()
		
		# Closing preferences window
		self.window.destroy()
		
		# Calling function to apply preferences on existing views
		self.apply_preferences()


	def on_preferences_dialog_restore(self, widget):
		self.set_default_preferences()
		self.apply_preferences()
		self.window.destroy()

	def on_preferences_dialog_close(self, widget):
		self.window.destroy()
